src/main/java/com/google/devtools/build/lib/analysis/PowershellCommandConstructor.java

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.analysis;
+
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.analysis.actions.FileWriteAction;
+
+/**
+ * The class for constructing command line for Powershell.
+ */
+public class PowershellCommandConstructor implements CommandConstructor {
+
+  // `powershell.exe` exists at C:\Windows\System32\WindowsPowerShell\v1.0, which is in the default
+  // PATH on Windows.
+  private static final String POWERSHELL_BIN = "powershell.exe";
+  private static final String[] POWERSHELL_SETUP_COMMANDS = {
```

Comment:
Could you store the joined string instead of the parts to join?
Also I believe we prefer `com.google.common.base.Joiner.on("; ").join(...)` to `String.join("; ", ...)`.